lib/preload.js

```diff
@@ -382,13 +382,56 @@ class Preloader extends EventEmitter {
 			.catch((error) => this.emit('error', error));
 	}
 
+	/**
+	 * Send a command to `preload.py` (running in a Docker container) through
+	 * its stdin, and read a response from its stdout. The stdout response
+	 * must be a JSON object containing a `statusCode` field (integer) and
+	 * optionally `result` and `error` fields. If `statusCode` is the number
+	 * zero, the command execution is considered successful, otherwise it is
+	 * assumed to have failed and the returned promise is rejected with an
+	 * error message including the message provided in the `error` field.
+	 *
+	 * @param {string} command
+	 * @param {object} parameters
+	 */
 	_runCommand(command, parameters) {
 		return new Bluebird((resolve, reject) => {
+			const cmd = JSON.stringify({ command, parameters }) + '\n';
 			this.stdout.once('error', reject);
 			this.stdout.once('data', (data) => {
-				resolve(JSON.parse(data).result);
+				let strData = data;
+				try {
+					strData = data.toString();
+				} catch (_e) {
+					// ignore
+				}
+				let response = {};
+				try {
+					response = JSON.parse(strData);
+				} catch (error) {
+					response.statusCode = 1;
+					response.error = error;
+				}
+				if (response.statusCode === 0) {
+					resolve(response.result);
+				} else {
+					const msg = [
+						`An error has occurred executing internal preload command '${command}':`,
+						cmd,
+					];
+					if (response.error) {
+						msg.push(
+							`Status code: ${response.statusCode}`,
```

Comment:
I was under the impression that something in the CLI was explicitly looking for `StatusCode` in a failed preload command, so this is the only place where we actually want our python graceful failure to mimic that of a dockerode failure. I might just not be clear on how we are using the reject errors.

Reply:
`statusCode` is not inspected anywhere else, only in this block of code. This error message is just a string that may be printed to console when the promise rejection bubbles up. It is just a coincidence that the name `statusCode` looks similar to `StatusCode` in the following lines:

```js
		this.container.wait().then((data) => {
				if (data.StatusCode !== 0) {
```

My motivation / rationale was to try and detect, around here, whether the Python script was successful in executing a "preload command" (JSON object written to the Python script's `stdin`). For this purpose, I insist on `response.statusCode === 0` being true (note the use of `===` instead of `==`, so I'm looking for "number zero" specifically, not any falsy value like empty string, null, undefined).